### ISSUE TITLE
[question] Help me choosing between Gorilla / RabbitMQ

### ISSUE THREAD

**Issue by bcashier** [gorilla/websocket#527]
Hi all,

I was searching for a solution that helps me implementing websockets on my project ... first I was thinking to use RabbitMQ but unfortunately not supporting do not pass autobahn test, sincerely I already integrated to my project a random one just to get some works done and not stucking on this.

Now that I have to get the better solution that feet my need, by the way my need is the PubSub system only ... I have to choose the fastest and low latency solution ... making a search on Google, I found out that you can handle 1Millions requests / second with only 1Go of RAM !

So I am really interested to know more about you and especially if using your solution will give better result than RabbitMQ ?

Another question, do Gorilla support multi threading ?

Thank you in advance for your help.

**Comment by elithrar** [comment]
Hi there -

It’s probably worth clarifying a few points:

• WebSockets are a protocol for streaming small messages over HTTP, but
provide no inherent “PubSub” capabilities. Buffering messages for clients,
replication, resending, authentication, fanning out to multiple subscribers
- that’s on you to implement on top.

• RabbitMQ is a message bus designed for persisting and sending events
between systems. A lot of the application-level semantics (authentication,
replication, etc) are built-in, with the caveat that it doesn’t work in a
browser environment (which is why WebSockets exist). RabbitMQ does not pass
the autobahn tests as those tests are for validating standards support of
*WebSockets*.

What are you trying to build? Are your clients browser-based, IoT devices,
or ?

**Comment by bcashier** [comment]
@elithrar Hi sir ! And thanks for your answer :)

My clients are browser based.

Clients subscribe to a specific topic from browser side using autobahn (javascript).

And publish message from the server to notify them about new changes, update datas on browser etc.

Waiting your answer and thanks in advance :)

**Comment by elithrar** [comment]
If your clients are browser-based, then you can’t (easily) use RabbitMQ.

WebSockets will give you a path to streaming data between clients and
server, but the concepts of “topics”, changes, data schema, etc - is up to
you.

The other option would be to use a cloud provider service like GCP’s PubSub
or AWS Kinesis.

**Comment by bcashier** [comment]
@elithrar 

Is it possible to implement the concept of Publish / Subscribe to a TOPIC using Gorilla ?

If yes, do you have an example of code already done on this way ?

Thank you :)

**Comment by elithrar** [comment]
Take a look at the chat example -
https://github.com/gorilla/websocket/tree/master/examples/chat - noting
that the concepts of rooms (topics, in your case) is up to you to implement.

**Comment by bcashier** [comment]
@elithrar Do the performances will be same even while using rooms ?

**Comment by elithrar** [comment]
I can't answer that - that's up to you to manage.

My recommendation is that you think about what you need to build first, and
what kind of scale it needs to achieve BEFORE focusing on performance
numbers. It sounds like you have a bit of learning to do (which is OK!)
around message queues & WebSockets before worrying about 1M+ clients.

**Comment by bcashier** [comment]
@elithrar Thank you for your answer :)

You are right about my knowledge on WebSockets, it's really limited ... Sincerely, I learned only what is needed for my project without taking in count the a close eye on how it works and even the basics.

My project need the concept of rooms since some details must be transmitted to a specific users after logged in (Sensitive informations) ... but others can be public and shared for the whole users/visitors.

> My recommendation is that you think about what you need to build first

I already used to finish the project but with poor performances, so I am looking to switch to a better languages with low latency and especially choosing the best solutions that works + gives better performances than the existing one, and especially question of scalability.

> worrying about 1M+ clients.

Worrying about +1M websockets sent per second is my goal actually... each day I am spending higher part of my time to gather the best existing open source technologies and building schema for a best combination of all modules etc.

Anyway many thanks for taking time to answer my question, appreciated.

**Comment by bcashier** [comment]
@elithrar Thank you for your answer :)

You are right about my knowledge on WebSockets, it's really limited ... Sincerely, I learned only what is needed for my project without taking in count the a close eye on how it works and even the basics.

My project need the concept of rooms since some details must be transmitted to a specific users after logged in (Sensitive informations) ... but others can be public and shared for the whole users/visitors.

> My recommendation is that you think about what you need to build first

I already used to finish the project but with poor performances, so I am looking to switch to a better languages with low latency and especially choosing the best solutions that works + gives better performances than the existing one, and especially question of scalability.

> worrying about 1M+ clients.

Worrying about +1M websockets sent per second is my goal actually and is the top priority before starting programming on it for a second time ... each day I am spending higher part of my time to gather the best existing open source technologies and building schema for a best combination of all modules etc.

Anyway many thanks for taking time to answer my question, appreciated.
